src/Raven.Server/Web/RequestHandlerContext.cs

```diff
@@ -17,7 +17,6 @@ public sealed class RequestHandlerContext
 
         public string DatabaseName => Database?.Name ?? DatabaseContext?.DatabaseName;
         public MetricCounters DatabaseMetrics => Database?.Metrics ?? DatabaseContext?.Metrics;
-
-        public string ClusterTransactionId => Database?.ClusterTransactionId;
+        public string ClusterTransactionId => Database?.ClusterTransactionId ?? DatabaseContext?.DatabaseRecord.GetClusterTransactionId();
```

Comment:
this is the actual fix, the rest is moving things around to support better testing abstraction